app/lib/cluster_viz_service.rb

```diff
@@ -59,10 +59,31 @@ def self.load_spatial_options(study)
     # now return an array of objects with names and associated cluster names
     spatial_file_info.map do |cluster|
       associated_cluster_names = cluster[1].map{ |id| associated_clusters[id] }
-      { name: cluster[0], associated_clusters: associated_cluster_names }
+      { name: cluster[0], associated_clusters: associated_cluster_names, bucket_file_name: cluster[2] }
     end
   end
 
+  def self.load_image_options(study)
+    # grab all the image files for this study
+    image_file_info = StudyFile.where(study: study, file_type: 'Image')
+                                 .pluck(:name, :spatial_cluster_associations, :upload_file_name, :description)
+
+    # now grab any non spatial cluster files for the study that had ids specified in the image files above
+    # and put them in an id=>name hash
+    associated_clusters = StudyFile.where(study: study, :id.in => image_file_info.map{ |si| si[1] }.flatten.uniq)
```

Comment:
True - it does.  If you want to find the middle ground, you could create hashes rather than arrays, so that we have key names to reference, rather than array indexes.  For instance:

```
attributes = [:name, :spatial_cluster_associations, :upload_file_name, :description]
image_file_info = StudyFile.where(study: study, file_type: 'Image')
                                .pluck(*attributes).map { |file_attr| Hash[attributes.zip(file_attr) } 
                                
```